zkvm/docs/zkvm-spec.md

```diff
@@ -1390,49 +1408,93 @@ Fails if:
+6. Adds an [output entry](#output-entry) to the [transaction log](#transaction-log).
 
 Immediate data `k` is encoded as [LE32](#le32).
 
+Fails if VM’s [last anchor](#vm-state) is missing.
+
 
 #### contract
 
 _items... pred_ **contract:_k_** → _contract_
 
 1. Pops [predicate](#predicate) `pred` from the stack.
 2. Pops `k` items from the stack.
-3. Creates a contract with the `k` items as a payload and the predicate.
+3. Creates a contract with the `k` items as a payload, the predicate `pred`, and anchor set to the [VM’s last anchor](#vm-state).
 4. Pushes the contract onto the stack.
+5. Update the [VM’s last anchor](#vm-state) with the [contract ID](#contract-id) of the new contract.
 
 Immediate data `k` is encoded as [LE32](#le32).
 
+Fails if VM’s [last anchor](#vm-state) is missing.
 
 #### nonce
 
-_predicate_ **nonce** → _contract_
+_pred blockid_ **nonce** → _contract_
+
+1. Pops [32-byte string](#data-type) `blockid` from the stack.
+2. Pops [predicate](#predicate) `pred` from the stack.
+3. Computes the nonce [anchor](#anchor) value:
+    ```
+    T = Transcript("ZkVM.nonce")
+    T.commit("blockid", blockid)
+    T.commit("predicate", pred)
+    T.commit("maxtime", LE64(tx.maxtime))
+    nonce_anchor = T.challenge_bytes()
+    ```
+4. Pushes a new [contract](#contract-type) with:
+    * an empty [payload](#contract-payload)
+    * predicate set to `pred`
+    * anchor set to `nonce_anchor`.
+5. Sets the [VM’s last anchor](#vm-state) to [contract ID](#contract-id) of that contract.
+6. Adds [nonce entry](#nonce-entry) to the [transaction log](#transaction-log) with the `blockid`, transaction [maxtime](#time-bounds) and `nonce_anchor`.
+
+Fails if:
+* `blockid` is a not a 32-byte [data string](#data-type).
+* `pred` is not a valid [point](#point).
+
+**Discussion**
+
+A nonce serves two purposes:
 
-1. Pops [predicate](#predicate) from the stack.
-2. Pushes a new [contract](#contract-type) with an empty [payload](#contract-payload) and this predicate to the stack.
-3. Adds [nonce entry](#nonce-entry) to the [transaction log](#transaction-log) with the predicate and transaction [maxtime](#time-bounds).
-4. Sets the [VM uniqueness flag](#vm-state) to `true`.
+1. It provides a unique anchor for an issuing transaction if other anchors are not available (i.e., from other values already on the stack);
+2. It binds the entire transaction to a particular blockchain, protecting not only against cross-blockchain replays, but also potential blockchain forks due to compromise of the old block-signing keys. This is a necessary feature for stateless signing devices that rely on blockchain proofs.
 
-Fails if `predicate` is not a valid [point](#point).
+Blockchain state machine performs the following checks:
+
+1. The `blockid` is checked against the IDs of “recent” blocks when the transaction is applied to the blockchain. It must match a recent block, or the initial block of the blockchain.
```

Comment:
TxVM also allows it to be 0x00 × 32 bytes. I'm fine with excluding that possibility but first let's make sure we remember why TxVM allows it in the first place. (I don't, do you?)

Reply:
It was allowed to support hypothetical case where we can do fancy replay of the same transaction on two TxVM chains to do an unspecified shorter cross-chain swap. However, having this possibility, precludes a guarantee that all anchors are globally unique. So if someone somewhere used an zero-bound nonce, its non-uniqueness can shoot someone's foot down the road through the chain of anchor derivation.

Since we are going to make "cross-chain swaps" AKA "payment channels" by the book and be happy with that, I don't think we need to keep this dangerous behavior.

Reply:
OK